finesse/hardware/plugins/temperature/senecat121.py

```diff
@@ -0,0 +1,148 @@
+
+from .temperature_monitor_base import TemperatureMonitorBase
+
+
+class SenecaError(Exception):
+    """Indicates that an error occurred while communicating with the device."""
+
+
+class SenecaT121(
+    SerialDevice,
+    TemperatureMonitorBase,
+    description="SenecaT121",
+    default_baudrate=57600,
+):
+    """An interface for Seneca T121 temperature readers.
+
+    This device communicates through the MODBUS-RTU protocol.
+
+    The manual for this device is available at:
+    https://www.seneca.it/products/k107usb/doc/installation_manualEN
+    """
+
+    def __init__(self, *serial_args: Any, **serial_kwargs: Any) -> None:
+        """Create a new Seneca from an existing serial device.
+
+        Args:
+            serial_args: Arguments to Serial constructor
+            serial_kwargs: Keyword arguments to Serial constructor
+        """
+        SerialDevice.__init__(self, *serial_args, **serial_kwargs)
+        TemperatureMonitorBase.__init__(self)
+
+    def close(self) -> None:
+        """Close the connection to the device."""
+        self.serial.close()
+
+    def read(self) -> bytes:
+        """Read temperature data from the Seneca.
+
+        Returns:
+            data: the sequence of bytes read from the device
+
+        Raises:
+            SenecaError: Malformed message received from device
+        """
+        try:
+            data = self.serial.read(size=21)
+        except SerialException as e:
+            raise SenecaError(e)
+
+        # require 21 bytes else checks will fail
+        min_length = 21
+        if len(data) != min_length:
+            raise SenecaError("Insufficient data read from device")
+
+        return data
+
+    def request_read(self) -> None:
+        """Write a message to the Seneca to prepare for a read operation.
+
+        A byte array of [1, 3, 0, 2, 0, 8, 229, 204] is written to the device as a
+        request to read the data. This byte array was taken from the original C# code.
+
+        Raises:
+            SenecaError: Error writing to the device
+        """
+        try:
+            self.serial.write(bytearray([1, 3, 0, 2, 0, 8, 229, 204]))
+        except Exception as e:
+            raise SenecaError(e)
+
+    def parse_data(self, data: bytes) -> list[Decimal]:
+        """Parse temperature data read from the Seneca.
+
+        The sequence of bytes is put through the conversion function and translated
+        into floats.
+
+        Args:
+            data: The bytes read from the device.
+
+        Returns:
+            vals: A list of Decimals containing the temperature values recorded
+                by the Seneca device.
+        """
+        dt = numpy.dtype(numpy.uint16).newbyteorder(">")
+        ints = numpy.frombuffer(data, dt, 8, 3)
+        print("Pre-scaling:", ints)
+
+        # Test for minimum and maximum output values
+        # ints = [4000, 20000]
+
+        vals = [Decimal(float(self.calc_temp(val))) for val in ints]
```

Comment:
Fair enough.

Ultimately, I think we want to change the API for `TemperatureMonitorBase`, because we don't *need* those values as a `list[Decimal]`. A numpy array of `np.float64`s would be fine too, so you could change the return type for all the `get_temperature()` functions to `Sequence[Decimal | np.float64]` or something. It might be better to alias this to `TemperatureSequence` or something. If you don't feel comfortable doing this, maybe open an issue for it?